core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +92,53 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    AtomicWaitAddressArgs *data = (AtomicWaitAddressArgs *)user_data;
+    if (data->len > 0) {
+        if (!(data->addr = wasm_runtime_realloc(data->addr,
+                                                data->len * sizeof(uint32)))) {
+            LOG_ERROR("failed to realloc atomic wait address list during "
+                      "wait_map traversal");
+            return;
+        }
+    }
+    else {
+        if (!(data->addr = wasm_runtime_malloc(sizeof(data->len) + 1))) {
+            LOG_ERROR("failed to malloc atomic wait address list during "
+                      "wait_map traversal");
+            return;
+        }
+    }
+    data->len++;
+    data->addr[data->len - 1] = key;
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    AtomicWaitAddressArgs *args = NULL;
+    if (!(args = wasm_runtime_malloc(sizeof(AtomicWaitAddressArgs)))) {
+        LOG_ERROR("failed to create atomic wait address args node");
+        return;
+    }
+
+    memset(args, 0, sizeof(*args));
+    os_mutex_lock(&shared_memory_list_lock);
+    // create list of addresses
+    bh_hash_map_traverse(wait_map, create_list_of_waiter_addresses, args);
+    os_mutex_unlock(&shared_memory_list_lock);
+
+    // notify
+    for (uint32 i = 0; i < args->len; i++) {
```

Comment:
@loganek do you think `len`'s value can become inconsistent due to a race condn? do you suggest putting an if check before calling notify i.e. `if(args->addr[i])`?